feat: ClusterProfile integration #24282

LogicalShark opened the pull request:
Closes https://github.com/argoproj/argo-cd/issues/24282

corentone commented on applicationset/controllers/clusterprofileeventhandler.go:
```diff
@@ -0,0 +1,170 @@
+package controllers
```
the logic in this file seems complex and I wonder if its a copy/paste from the existing. in that case, any chance the existing logic could be edited (made polymorphic?) to allow for ClusterProfile and avoid the duplication?
I'm afraid otherwise maintaining of the two would be hard.
If the plan is to move fully to ClusterProfile (ambitious but possible), I think this file kept as is is fine and we should review it properly.

corentone commented on docs/user-guide/auto_sync.md:
````diff
@@ -111,6 +111,23 @@ spec:
       refresh: true
 ```
 
+## Automatic Retry Refresh on new revisions
````
why is this part of this PR?

corentone commented on applicationset/generators/cluster_profile.go:
```diff
@@ -0,0 +1,251 @@
+package generators
```
this file still seems to have a lot of references to secrets, which may be because you forked it from the Secrets-based generator. I'm guessing it's still a WIP, could you update us when it's ready?

If it's ready for review, Ill have a closer look but it sounds like there may be a few references to secrets we could get rid of

LogicalShark committed:
Secret syncer

Signed-off-by: LogicalShark <maralder@google.com>

FourFifthsCode commented on docs/user-guide/cluster-profiles-gcp-example.md:
```diff
@@ -0,0 +1,212 @@
+              mountPath: /app/cp-creds
+          args:
+            - "/usr/local/bin/argocd-applicationset-controller"
+            - "--cluster-profile-providers-file=/app/cp-creds/cp-creds.json"'
```
Would be nice if we could bake some of this into some default providers or make it a CRD or something so we wouldn't have to patch the argo manifests.

mikeshng committed:
Add ExecProvider.Config support for exec credential plugins

Pass cluster-specific configuration (e.g., cluster name) to exec credential plugins via ExecCredential.Spec.Cluster.Config. Required for multi-cluster authentication plugins that need per-cluster data.

Signed-off-by: Mike Ng <ming@redhat.com>

LogicalShark committed:
Support built-in cloud providers without requiring mounted access providers file

Signed-off-by: LogicalShark <maralder@google.com>

LogicalShark committed:
Move clusterprofile syncer to separate controller

Signed-off-by: LogicalShark <maralder@google.com>

crenshaw-dev commented:
Thanks for all your effort on this @LogicalShark! At the moment I think the maintainers have their plate full with other work. I don't want this to stall, but I don't think we can dedicate adequate time to it at the moment.

I'd recommend [proposing](https://github.com/argoproj/argoproj/blob/main/community/ecosystem-projects.md#project-proposal-issue) adding the feature as a new argoproj-labs project. There you'll be able to iterate quickly without needing Argo maintainers to approve each PR. Once the project has stabilized and gained some adoption, we can assess moving it in-tree.

LogicalShark commented:
Implemented in labs https://github.com/argoproj-labs/clusterprofile-integration-for-argocd! If it is decided in the future to integrate into argo-cd, we can make a new PR.
